src/opt/memory_coloring_impl.hpp

```diff
@@ -145,7 +151,54 @@ struct memory_coloring_impl
             return (i1->offset > i2->offset);
         }
     };
-    module* p_program;
+
+    using map_deps = std::unordered_map<instruction_ref, std::unordered_set<instruction_ref>>;
+    void
+    calc_implicit_deps(const module& smod, const module& pmod, instruction_ref ins, map_deps& deps)
+    {
+        const auto& ins_inputs = ins->inputs();
+        for(auto ii : iterator_for(smod))
+        {
+            const auto& ii_inputs = ii->inputs();
+            for(auto iii : ii_inputs)
+            {
+                if(pmod.has_instruction(iii))
+                {
+                    if(not contains(ins_inputs, iii))
+                        deps[ins].insert(iii);
+                }
+            }
+
+            const auto& mod_args = ii->module_inputs();
+            if(not mod_args.empty())
+            {
+                for(const auto* ssmod : mod_args)
+                {
+                    calc_implicit_deps(*ssmod, pmod, ins, deps);
+                }
+            }
+        }
+    }
+
+    void calc_implicit_deps(module& p)
```

Comment:
Doesnt the scheduler need this function? Also this should be made a member function to the module class and it should also return the map.

Reply:
moved to module as a member function